docs/users_guide/exts/qualified_do.rst

```diff
@@ -0,0 +1,167 @@
+.. _qualified-do-notation:
+
+Qualified do-notation
+-------------------------
+
+.. index::
+   single: Qualified do-notation
+   single: do-notation; Qualified
+
+.. extension:: QualifiedDo
+    :shortdesc: Enable qualified do-notation desugaring.
+
+    :since: 8.12.1
+
+    Allow the use of qualified ``do`` notation.
+
+``QualifiedDo`` enables qualifying a ``do`` block with a module name, to control which operations to use for
+the monadic combinators that the ``do`` notation desugars to.
+
+With ``-XQualifiedDo`` is enabled, you can *qualify* the ``do`` notation by writing ``modid.do``, where
+``modid`` is a module name in scope: ::
+
+    {-# LANGAUGE QualifiedDo #-}
+    import qualified Some.Module.Monad as M
+
+    action :: M.SomeType a
+    action = M.do x <- u
+                  res
+                  M.return x
+
+
+The additional module name (here ``M``) is called the qualifier of the do-expression.
+
+The semantics of ``do`` notation statements with ``-XQualfiedDo`` is as follows:
+
+* The ``x <- u`` statement uses ``(M.>>=)`` ::
+
+    M.do { x <- u; stmts }  =  u M.>>= \x -> M.do { stmts }
+
+* The ``u`` statement uses ``(M.>>)`` ::
+
+    M.do { u; stmts }  =  u M.>> M.do { stmts }
+
+* The a ``pat <- u`` statement uses ``M.fail`` for the failing case,
+  if such a case is needed ::
+
+    M.do { pat <- u; stmts }  =  u M.>>= \case
+      { pat -> M.do { stmts }
+      ; _ -> M.fail "…"
+      }
+
+  If the pattern cannot fail, then we don't need to use ``M.fail``.  ::
+
+    M.do { pat <- u; stmts }  =  u M.>>= \case pat -> M.do { stmts }
+
+*  The desugaring of ``-XApplicativeDo`` uses ``(M.<$>)``, ``(M.<*>)``,
+   and ``M.join`` (after the the applicative-do grouping has been performed) ::
+
+    M.do { (x1 <- u1 | … | xn <- un); M.return e }  =
+      (\x1 … xn -> e) M.<$> u1 M.<*> … M.<*> un
+
+    M.do { (x1 <- u1 | … | xn <- un); stmts }  =
+      M.join ((\x1 … xn -> M.do { stmts }) M.<$> u1 M.<*> … M.<*> un)
+
+
+  Note that ``M.join`` is only needed if the final expression is
+  not identifiably a ``return``.
+
+*  With ``-XRecursiveDo``, ``rec`` blocks use ``M.mfix`` and ``M.return``: ::
+
+     M.do { rec { x1 <- u1; … ; xn <- un }; stmts }  =
+       M.do
+       { (x1, …, xn) <- M.mfix (\~(x1, …, xn) -> M.do { x1 <- u1; …; xn <- un; M.return (x1, …, xn)})
+       ; stmts
+       }
+
+If a name ``M.op`` is required by the desugaring process (and only if it's required!) but the name is
+not in scope, an error message is produced:
+
+  “Desugaring statement ``<stmt>`` requires ``<M.op>`` which is not in scope”
```

Comment:
```suggestion
not in scope, it is reported as an error.
```

The error text does not match closely the implementation, and it is prone to get outdated if we try to copy it.